File: event-bus-rabbit/src/main/java/pegasus/eventbus/topology/FallbackTopologyManager.java
package pegasus.eventbus.topology;

import java.util.concurrent.TimeoutException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import pegasus.eventbus.amqp.RoutingInfo;
import pegasus.eventbus.amqp.TopologyManager;
import pegasus.eventbus.client.EventManager;
import pegasus.eventbus.topology.event.EventTypeRoutingInfo;
import pegasus.eventbus.topology.event.GetEventTypeRoute;

//TODO: PEGA-720 This class need test coverage.
public class FallbackTopologyManager implements TopologyManager {

	   protected static final Logger LOG              = LoggerFactory.getLogger(FallbackTopologyManager.class);

	   private EventManager          eventManager;

	@Override
	public RoutingInfo getRoutingInfoForEvent(Class<?> eventType) {
		GetEventTypeRoute request = new GetEventTypeRoute(eventType.getCanonicalName());
		try {
			@SuppressWarnings("unchecked")
			EventTypeRoutingInfo response = eventManager.getResponseTo(request, 1000, EventTypeRoutingInfo.class);
			return response.getRouteInfo();
		} catch (InterruptedException e) {
			LOG.warn("Thread interrupted while waiting for route info for event type: " + eventType.getCanonicalName(), e);
		} catch (TimeoutException e) {
			LOG.warn("Timed out while waiting for route info for event type: " + eventType.getCanonicalName(), e);
		}
		return null;
	}

	@Override
	public RoutingInfo[] getRoutingInfoForNamedEventSet(String eventSetName) {
		return null;
	}

	@Override
	public void start(EventManager eventManager) {

        LOG.trace("Fallback Topology Service Manager starting.");

        this.eventManager = eventManager;
	}

	@Override
	public void close() {

	}

}

File: event-bus-rabbit/src/main/dotnet/topology/FallbackTopologyManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using pegasus.eventbus.client;
using pegasus.eventbus.amqp;


namespace pegasus.eventbus.topology
{
    public class FallbackTopologyManager : ITopologyService
    {
        private IEventManager _eventMgr;


        public void Start(IEventManager eventManager)
        {
            _eventMgr = eventManager;
        }

        public void Close()
        {
            throw new NotImplementedException();
        }

        public RoutingInfo GetRoutingInfoForEventOfType(Type evType)
        {
            GetEventTypeRoute request = new GetEventTypeRoute(evType.FullName);
            RoutingInfo route = null;

            try
            {
                EventTypeRoutingInfo response = _eventMgr.GetResponseTo<EventTypeRoutingInfo>(request, 5 * 1000);
                route = response.RouteInfo;
            }
            catch (Exception ex)
            {

            }

            return route;
        }

        public IEnumerable<RoutingInfo> GetRoutingInfoForNamedEventSet(string eventSetName)
        {
            return null;
        }
    }
}
